backend/app/src/main/kotlin/io/tolgee/api/v2/controllers/UserMfaController.kt

```diff
@@ -0,0 +1,50 @@
+package io.tolgee.api.v2.controllers
+
+import io.swagger.v3.oas.annotations.Operation
+import io.swagger.v3.oas.annotations.tags.Tag
+import io.tolgee.dtos.request.UserMfaRecoveryRequestDto
+import io.tolgee.dtos.request.UserTotpDisableRequestDto
+import io.tolgee.dtos.request.UserTotpEnableRequestDto
+import io.tolgee.security.AuthenticationFacade
+import io.tolgee.security.JwtTokenProvider
+import io.tolgee.security.patAuth.DenyPatAccess
+import io.tolgee.security.payload.JwtAuthenticationResponse
+import io.tolgee.service.MfaService
+import org.springframework.web.bind.annotation.*
+import javax.validation.Valid
+
+@RestController
+@RequestMapping("/v2/user/mfa")
+@Tag(name = "User Multi-Factor Authentication")
+class UserMfaController(
+  private val authenticationFacade: AuthenticationFacade,
+  private val mfaService: MfaService,
+  private val jwtTokenProvider: JwtTokenProvider
+) {
+  @PutMapping("/totp")
+  @Operation(summary = "Enables TOTP-based two-factor authentication. Invalidates all previous sessions upon success.")
+  @DenyPatAccess
+  fun enableMfa(@RequestBody @Valid dto: UserTotpEnableRequestDto): JwtAuthenticationResponse {
+    mfaService.enableTotpFor(authenticationFacade.userAccountEntity, dto)
+    return JwtAuthenticationResponse(
+      jwtTokenProvider.generateToken(authenticationFacade.userAccountEntity.id).toString()
+    )
+  }
+
+  @DeleteMapping("/totp")
+  @Operation(summary = "Disables TOTP-based two-factor authentication. Invalidates all previous sessions upon success.")
+  @DenyPatAccess
+  fun disableMfa(@RequestBody @Valid dto: UserTotpDisableRequestDto): JwtAuthenticationResponse {
+    mfaService.disableTotpFor(authenticationFacade.userAccountEntity, dto)
+    return JwtAuthenticationResponse(
+      jwtTokenProvider.generateToken(authenticationFacade.userAccountEntity.id).toString()
```

Comment:
Why new token?

Reply:
> Sessions are invalidated upon changing password or enabling/disabling 2FA

Enabling or disabling MFA causes sessions to be invalidated, so a new token is returned to keep the current session alive

Reply:
I don't understand. Where it causes it?

Reply:
https://github.com/tolgee/tolgee-platform/blob/73983eba9c7eaf60a9e969c123d69f464ea6f7c1/backend/data/src/main/kotlin/io/tolgee/service/UserAccountService.kt#L167

It's an intentional behavior I've seen from multiple platforms in my experience, as changing 2FA settings is considered credentials change
it's easy to drop this behavior if it's something you'd prefer not to happen

Reply:
Aha, OK, I was only curious, how you can invalidate JWT token. But this way seems cool!